src/messaging/ExchangeContext.cpp

```diff
@@ -489,6 +490,9 @@ void ExchangeContext::NotifyResponseTimeout(bool aCloseIfNeeded)
     }
 #endif // CONFIG_DEVICE_LAYER && CHIP_CONFIG_ENABLE_ICD_SERVER
 
+    // Grab the value of WaitingForResponseOrAck() before we mess with our state.
+    bool gotMRPAck = !WaitingForResponseOrAck();
```

Comment:
There doesn't seem any difference between the flags WaitingForResponseOrAck and WaitingForAck in the way they are actually used. So, is the former required? We could use the latter for gotMRPAck instead, no?
Because once we receive the application response, we would be implicitly treating it as an Ack for the sent message.

Reply:
> There doesn't seem any difference between the flags WaitingForResponseOrAck and WaitingForAck in the way they are actually used

There is absolutely a difference.  `WaitingForResponseOrAck` gets set to false when we get a response (which might carry an ack... but we might be able to tell that) or an ack.  `WaitingForAck` gets set to false when we get an ack or stop listening for acks.

> Because once we receive the application response, we would be implicitly treating it as an Ack for the sent message.

The point is: we don't.  And the reason we don't is that once `WaitingForAck` goes false we are no longer tracking what message id has an outstanding ack (if any), and hence can't tell whether to treat the application response as that ack...

We could consider changing that, having a single `WaitingForAck` bit that stays true of MRP runs out of resends, and treating an app response, no matter whether it carries an ack or not (?), as an ack if we are no longer in the MRP table.  That involves some more widespread logic changes than what we did here, but would be reasonably equivalent.

Reply:
From what I read, it gets set to false in ReliableMessageContext.cpp:108, rt? That is when MRP is processing its Ack, be it either a StandaloneAck or a piggybacked Ack. I was treating them both as the same, i.e, an acknowledgment of delivery of the sent message, hence an Ack.
Yes, once WaitingForAck goes false, we have removed the message from RetransmitTable and do not care about retransmission anymore. However, Acks can still arrive after that, as you stated(maybe in a piggybacked form). These are duplicate Acks and are generally discarded.

Reply:
> it gets set to false in ReliableMessageContext.cpp:108, rt?

And also in all the places where `SetResponseExpected(false)` is called.

> These are duplicate Acks and are generally discarded.

The point is: if we removed fthe message from RetransmitTable because we ran out of retransmits, they may not in fact be duplicate acks.  But we just can't tell, because we removed our state that would enable us to tell.

Reply:
I thought we were removing from the retransmit table only on all MRP retries timeout or an Ack was received. Are we evicting an outstanding retransmit to make room for a new Send? That should be a SendError because our RetransmitTable is full, no?

Reply:
> I thought we were removing from the retransmit table only on all MRP retries timeout or an Ack was received. 

That is correct. The question is what happens when a message is received after the MRP retries timeout but before response timeout.

Reply:
From an MRP standpoint, this message means nothing as all state for Ack receipt is gone(although this message is proof of delivery of the sent message). However, since Exchange has not timed out, the application may be waiting for the response still, and the question is should this be delivered up to the app? Or thrown away. I guess, if the ExchangeContext exists, then it MAY be sent up to the app bypassing all MRP checks. 
One option would have been to notify the App with OnSendError() upon MRP timeout, and that would have allowed the App to either close the exchange or wait. But in the absence of that, the App will wait until ResponseTimeout and if the Response comes in within the timeout and the Exchange is available, then it can be sent to the app. WDYT?

Reply:
>  this message means nothing as all state for Ack receipt is gone

Yep, that is what we implement.

> the question is should this be delivered up to the app

The answer to this question is "yes".  ;)

> Or thrown away.

It being thrown away (due to recent changes to the MRP logic which this PR undoes) is _exactly_ what this PR is fixing.  Throwing it away leads to an unacceptably high rate of commissioning failures.

> WDYT?

Sounds like we are in agreement.